doc/arch/adr-002-geoprocessing-caching.md

```diff
@@ -0,0 +1,103 @@
+
+Outside of geoprocessing, there are also [some](https://github.com/WikiWatershed/model-my-watershed/blob/dea12ff9a6fb234e30978e97ed1d0f6266a406c9/src/mmw/apps/modeling/calcs.py#L28) [database](https://github.com/WikiWatershed/model-my-watershed/blob/dea12ff9a6fb234e30978e97ed1d0f6266a406c9/src/mmw/apps/modeling/calcs.py#L51) [calls](https://github.com/WikiWatershed/model-my-watershed/blob/dea12ff9a6fb234e30978e97ed1d0f6266a406c9/src/mmw/apps/modeling/calcs.py#L97), but these are much faster in comparison and are not likely to see significant improvement by caching, especially since the cache could be a table in the database itself.
+
+In this ADR we consider the following questions:
+
+ * What will be cached?
+ * Where will it be cached?
+ * When will it be cached?
+ * How will items be identified, cached, and retrieved?
+ * How will the cache be invalidated?
+
+We also sketch out an implementation plan, and consider consequences and side-effects.
+
+## Decisions
+
+### What will be cached?
+
+The output of the geoprocessing service for WKAoIs will be cached. This is almost always a JSON blob of key-value pairs, where the key is a combination of overlaid cell values in a set of rasters, and the value is a count of such cells. The inputs are a GeoJSON shape and a [set of related arguments](https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/mmw/settings/base.py#L412) that specify rasters, operation type, CRS, etc.
+
+While we could cache the entire output of a MapShed or TR-55 run, in the case of an update to some of the constituent raster or vector data would force recalculation of the whole. By caching only the time-taking geoprocessing results, we ensure that any updates to constituent rasters would invalidate only that specific cached result, leaving the others still current. And since vector data results would never be cached, they will always be current upon update as well.
+
+In case we do not foresee updating the rasters very often, it may be beneficial to consider caching the entire output.
+
+### Where will it be cached?
+
+The current app stack gives us two options to choose as a cache: Redis and PostgreSQL. While reads from Redis can potentially be a bit faster, the in-memory cache will be purged upon restart and deployment. Furthermore, the use case to optimize is a faster read by one user over time, rather than parallel reads by many users concurrently, the speed advantages of Redis are of little use.
+
+By storing the cached results in the database, we ensure they are valid and available after deployments and new releases, while still being considerably faster than before.
+
+### When will it be cached?
+
+When a user selects a WKAoI, and a request is made to `/analyze` or `/modeling`, if the WKAoI results haven't already been cached, we will run and cache them. Over time, we will build up the cache, so that when new users request the same WKAoIs, they will get the cached results.
```

Comment:
We should think about the case where corrupt/invalid data is cached which causes an analyze job to fail repeatedly.